src/deepsparse/server/monitoring.py

```diff
@@ -0,0 +1,151 @@
+
+import pydantic
+import requests
+import yaml
+
+from deepsparse.server.config import ServerConfig, _endpoint_diff
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def start_file_watcher(
+    config_path: str, endpoints_url: str, check_interval_s: float
+) -> multiprocessing.Process:
+    """
+    Creates and starts a separate process that monitors `config_path`
+    for changes every `check_interval_s` seconds.
+    If endpoints are modified it sends requests to `endpoints_url`
+    to update the server.
+
+    :param config_path: Path to file to watch
+    :param endpoints_url: Url to send requests to for updating
+    :param check_interval_s: Seconds between each check.
+    :return: A **started** daemon `Process`.
+    """
+
+    proc = multiprocessing.Process(
+        target=_watch_file,
+        kwargs=dict(
+            config_path=config_path,
+            endpoints_url=endpoints_url,
+            check_interval_s=check_interval_s,
+        ),
+        daemon=True,
+    )
+    proc.start()
+    return proc
+
+
+def _watch_file(config_path: str, endpoints_url: str, check_interval_s: float):
+    for _ in _diff_generator(config_path, endpoints_url, check_interval_s):
+        ...
+
+
+def _diff_generator(
+    config_path: str, endpoints_url: str, check_interval_s: float
+) -> Generator[Optional[Tuple[ServerConfig, ServerConfig, str]], None, None]:
+    """
+    A generator that does the following:
+    1. Monitors for changes on `config_path`. It checks every `check_interval_s`.
+    2. If content has not changed, yields None
+    3. If content has changed, sends requests to server to update the config,
+       and yields the old and new server. It also saves the old verison of the config
+       with `.v{version}` postfixed to the path, where version is incremented after
+       every content change.
+    """
+    content = _ContentMonitor(config_path)
+
+    version = 0
+
+    while True:
+        time.sleep(check_interval_s)
+
+        diff = content.maybe_update_content()
+        if diff is None:
+            yield None
+            continue
+
+        old_content, new_content = diff
+        _LOGGER.info(f"Detected change in {config_path}")
+
+        try:
+            old_config = ServerConfig(**yaml.safe_load(old_content))
+            new_config = ServerConfig(**yaml.safe_load(new_content))
+        except yaml.error.YAMLError:
+            _LOGGER.error("Failed to read yaml, not updating.", exc_info=1)
+            yield None
+            continue
+        except pydantic.ValidationError:
+            _LOGGER.error("Unable to load ServerConfig, not updating.", exc_info=1)
+            yield None
+            continue
+
+        try:
+            _update_endpoints(endpoints_url, old_config, new_config)
+        except requests.RequestException:
+            yield None
+            _LOGGER.error("Requests to server failed, not updating.", exc_info=1)
+            continue
+
+        path = config_path + f".v{version}"
```

Comment:
Thoughts on saving previous versions of into a subdir next to the active config file, named "previous_configs" or something of that nature. Also, I think adding the date and time of modification to the config file name or as a markdown comment at the top of the config itself would be useful to users

Reply:
Like both of those, will do that!

Reply:
Added `<config path>.versions/` and a comment at the top of the file